core/primitives/src/time.rs

```diff
@@ -1,155 +1,316 @@
-            } else {
-                panic!("Use MockClockGuard in your test");
-            }
-        });
+    impl BorshSerialize for Time {
+        fn serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
+            let nanos = self.to_unix_timestamp() as u64;
+            BorshSerialize::serialize(&nanos, writer).unwrap();
+            Ok(())
+        }
     }
 
-    pub fn add_instant(mock_instant: Duration) {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                clock.durations.push_back(mock_instant);
-            } else {
-                panic!("Use MockClockGuard in your test");
-            }
-        });
+    impl BorshDeserialize for Time {
+        fn deserialize(buf: &mut &[u8]) -> Result<Self, std::io::Error> {
+            let nanos: u64 = borsh::BorshDeserialize::deserialize(buf)?;
+
+            Ok(Time::from_unix_timestamp(nanos))
+        }
     }
 
-    pub fn utc() -> DateTime<chrono::Utc> {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                let x = clock.pop_utc();
-                match x {
-                    Some(t) => t,
-                    None => {
-                        panic!("Mock clock run out of samples");
-                    }
-                }
-            } else {
-                chrono::Utc::now()
-            }
-        })
+    impl From<SystemTime> for Time {
+        fn from(system_time: SystemTime) -> Self {
+            Self { system_time }
+        }
     }
 
-    pub fn instant() -> Instant {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                let x = clock.pop_instant();
-                match x {
-                    Some(t) => t,
-                    None => {
-                        panic!("Mock clock run out of samples");
-                    }
-                }
-            } else {
-                Instant::now()
-            }
-        })
+    impl From<DateTime<Utc>> for Time {
+        fn from(utc: DateTime<Utc>) -> Self {
+            // utc.timestamp_nanos() returns i64
+            let nanos = utc.timestamp_nanos() as u64;
+
+            Self::UNIX_EPOCH + Duration::from_nanos(nanos)
+        }
     }
 
-    pub fn instant_call_count() -> u64 {
-        MockClockPerThread::with(|clock| clock.instant_call_count)
+    impl Time {
+        pub const UNIX_EPOCH: Time = Time { system_time: SystemTime::UNIX_EPOCH };
+
+        pub fn now() -> Self {
+            Self::UNIX_EPOCH + Duration::from_nanos(Clock::utc().timestamp_nanos() as u64)
+        }
+
+        pub fn duration_since(&self, rhs: &Self) -> Duration {
+            self.system_time.duration_since(rhs.system_time).unwrap_or(Duration::from_millis(0))
+        }
+
+        pub fn saturating_duration_since(&self, rhs: &Self) -> Duration {
+            self.system_time.duration_since(rhs.system_time).unwrap_or(Duration::from_millis(0))
+        }
```

Comment:
The methods are identical, seems better to have just `duration_since` and document that all ops are saturating